core/src/main/java/org/apache/carbondata/core/datamap/status/DataMapStatusManager.java

```diff
@@ -122,9 +136,99 @@ public static void dropDataMap(String dataMapName) throws IOException, NoSuchDat
+   * operations on main table
+   *
+   * @param allDataMapSchemas of main carbon table
+   * @throws IOException
+   */
+  public static void truncateDataMap(List<DataMapSchema> allDataMapSchemas) throws IOException {
+    for (DataMapSchema datamapschema : allDataMapSchemas) {
+      if (null != datamapschema.getRelationIdentifier() && !datamapschema.isIndexDataMap()) {
+        RelationIdentifier dataMapRelationIdentifier = datamapschema.getRelationIdentifier();
+        SegmentStatusManager segmentStatusManager = new SegmentStatusManager(AbsoluteTableIdentifier
+            .from(dataMapRelationIdentifier.getTablePath(),
+                dataMapRelationIdentifier.getDatabaseName(),
+                dataMapRelationIdentifier.getTableName()));
+        ICarbonLock carbonLock = segmentStatusManager.getTableStatusLock();
+        try {
+          if (carbonLock.lockWithRetries()) {
+            LOGGER.info(
+                "Acquired lock for table" + dataMapRelationIdentifier.getDatabaseName() + "."
+                    + dataMapRelationIdentifier.getTableName() + " for table status updation");
+            String metaDataPath =
+                CarbonTablePath.getMetadataPath(dataMapRelationIdentifier.getTablePath());
+            LoadMetadataDetails[] loadMetadataDetails =
+                SegmentStatusManager.readLoadMetadata(metaDataPath);
+            for (LoadMetadataDetails entry : loadMetadataDetails) {
+              entry.setSegmentStatus(SegmentStatus.MARKED_FOR_DELETE);
+            }
+            SegmentStatusManager.writeLoadDetailsIntoFile(
+                CarbonTablePath.getTableStatusFilePath(dataMapRelationIdentifier.getTablePath()),
+                loadMetadataDetails);
+          } else {
+            LOGGER.error("Not able to acquire the lock for Table status updation for table "
+                + dataMapRelationIdentifier.getDatabaseName() + "." + dataMapRelationIdentifier
+                .getTableName());
+          }
+        } finally {
+          if (carbonLock.unlock()) {
+            LOGGER.info("Table unlocked successfully after table status updation"
+                + dataMapRelationIdentifier.getDatabaseName() + "." + dataMapRelationIdentifier
+                .getTableName());
+          } else {
+            LOGGER.error("Unable to unlock Table lock for table" + dataMapRelationIdentifier
+                .getDatabaseName() + "." + dataMapRelationIdentifier.getTableName()
+                + " during table status updation");
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * This method checks if main table and datamap table are synchronised or not. If synchronised
+   * return true to enable the datamap
+   * @param dataMapSchema of datamap to be disabled or enabled
+   * @return flag to enable or disable datamap
+   * @throws IOException
+   */
+  public static boolean canDataMapBeEnabled(DataMapSchema dataMapSchema) throws IOException {
```

Comment:
This needs to be taken inside the lock of datamap status otherwise in concurrent scenarios one might enable the datamap which is intended to be disabled. So it should be in the same transaction.